src/client/datascience/data-viewing/jupyterVariableDataProvider.ts

```diff
@@ -62,41 +78,58 @@ export class JupyterVariableDataProvider implements IJupyterVariableDataProvider
         this.variable = variable;
     }
 
-    public async getDataFrameInfo(): Promise<IDataFrameInfo> {
+    public async getDataFrameInfo(sliceExpression?: string): Promise<IDataFrameInfo> {
         let dataFrameInfo: IDataFrameInfo = {};
         await this.ensureInitialized();
-        if (this.variable) {
+        let variable = this.variable;
+        if (variable) {
+            if (sliceExpression) {
+                variable = await this.variableManager.getDataFrameInfo(variable, this.notebook, sliceExpression);
```

Comment:
This seems weird? You're caching the variable. At the very least shouldn't it be called something else, like the 'slicedVariable' or something?

Is there the potential for some other code elsewhere not to understand which slice it is looking at?

Reply:
Yeah later on the variable name is used to make the row request. If the variable changes here, wouldn't that affect the stuff later? 

What is the reason to refetch here? Does it get a new shape? Perhaps you should just be caching the returned shape with something like 'currentSliceShape' or something?

Reply:
So really you're just refetching the shape? Should that be the newly cached value?

Reply:
No I'm refetching everything. New shape, new rowCount, etc

Reply:
I think the point of confusion might be over the fact that in `IJupyterVariable`, the `name` property never changes, but the `sliceExpression` property does. All we need to resolve a new variable is a `name` and, if provided, a `sliceExpression`. So when asking for variable info or rows based on the name property we are always guaranteed to recover the original variable.

Reply:
I think the problem I have is that there's a two step process and we're caching the second step in the same place as the first. 

First step - get data about the entire data frame (although I guess this would not technically be a dataframe anymore? Cause it might have more than 2 dimensions). Save this information in the this.variable field. I'm assuming this has all of the dimensions and what the UI will use to build the slicing control.

Second step - get data about a specific slice. This will be 2D data. It's being stored in this.variable field as well.

These seem like they're not really related? That maybe there should be two different messages for them? And they should be cached individually?

And for the second step, it should be cached based on the original variable name and the slice expression?

Reply:
The reason I'm caching everything in one spot is that I don't see a reason to keep the original variable around. The user asks to see a variable, we fetch the whole variable. The jupyterVariableDataProvider will never be asked for the original variable again, unless the user asks to see the original variable, but the original variable is just one type of slice. We will only ever show the user one slice at a time, so I didn't want to start caching maps of slice expressions to data (although I can also do that if you prefer).

What I have here looks weird because it will result in us always refetching when any sliceExpression is passed, even if the currently cached variable represents the same slice as the caller requested. However this isn't actually an issue because I'm also caching the sliceExpression and the variable together in dataViewer.ts (previously we were already caching the variable), so by the time we actually ask the jupyterVariableDataProvider for info, the slice must have changed.

In any case, would you prefer that I mirror the caching behavior in dataViewer.ts here?

```ts
if (!this.dataFrameInfoPromise || sliceExpression !== this.currentSlice) {
```

Reply:
> The jupyterVariableDataProvider will never be asked for the original variable again, unless the user asks to see the original variable, but the original variable is just one type of slice. We will only ever show the user one slice at a time, so I didn't want to start caching maps of slice expressions to data (although I can also do that if you prefer).

This is implicit, right? There's nothing in the code you're changing here that guarantees this?

I'm not saying your code won't work, I'm worried about the implicit assumptions about behavior that are in the code.

Reply:
If we can either prevent a second call for the whole enchilada (in this code isntead of because of how the UI works) or make a separate function that's called to make it super explicit, I'd rather have that.

Reply:
Yeah that's fair, nothing guarantees this. 

So you'd prefer a `this.variable`, `this.variableSlice` and a `this.variableSliceExpression`?

Reply:
I guess I'd prefer two API calls. They both essentially do the same thing but they're obvious they're for different use cases. Under the covers they call the same python code we wrote but in this class there's a two step process. Then we could also throw an exception if the first API is called more than once.

But if you don't agree, feel free to not do it that way. It's really just a way to make the code seem more specific to me. If you don't think the implicit thing is a big deal, that's fine with me.

Reply:
I think this was exactly the question I had earlier this week--whether to add a set of new methods or to extend the functionality of existing methods with an optional parameter. I actually [originally implemented](https://github.com/microsoft/vscode-jupyter/pull/4402/commits/d6a250450395f690cad0af018a21ad3fa2e41121) it the way you describe, but thought better of it because I was hesitant to introduce a new set of methods down the call hierarchy when the method implementations would be virtually identical to something that already exists. 

Just to be clear, it sounds like the preference is for the former. Something like 

```ts
getSlice?(variable: IJupyterVariable, sliceExpression: string): Promise<IDataFrameInfo> // Optional because dataViewer.ts consumes IDataViewerDataProvider, so we need to define this method on IDataViewerDataProvider and not on IJupyterVariableDataProvider 
```

And this would then call into `this.variableManager.getDataFrameInfo`? Or would you prefer a separate `this.variableManager.getSlice` as well?

Reply:
All the way down to the spot where we call the python code. If we weren't caching it wouldn't be as big a deal. But there's logic that's dependent upon the outside world not calling something in the variableManager too.

Reply:
So debuggerVariables.ts, kernelVariables.ts, jupyterVariables.ts, jupyterVariableDataProvider.ts/IDataViewerDataProvider are each going to have the following methods added:

```ts
getSliceInfo(
        targetVariable: IJupyterVariable,
        notebook?: INotebook,
        sliceExpression?: string
): Promise<IJupyterVariable>;
getSliceRows(
        targetVariable: IJupyterVariable,
        start: number,
        end: number,
        notebook?: INotebook,
        sliceExpression?: string
): Promise<JSONObject>;
```

And their definitions are going to basically be the same as getDataFrameInfo and getDataFrameRows? That feels kinda wasteful to me, but I'll make that change if you think it's more readable.

Reply:
Maybe we should just chat.

Reply:
I think if you got rid of the caching or cached based on slice expression, there would be no need for extra functions

Reply:
I guess they could share private `getExpressionInfo` and `getExpressionRows` impl methods, which would be less duplicative... But isn't your original concern about caching the slice in a spot that looks like the variable? Wouldn't another solution be to simply never cache an IJupyterVariable representing a slice, or, as we discussed, cache it as a separate private property? I guess adding a new set of methods feels like overkill for the cache problem.

Reply:
Ugh Github. Just saw your reply. Yeah I'll change it to cache based on slice expression.